addons/website/static/src/components/configurator/configurator.js

```diff
@@ -179,23 +179,39 @@ class DescriptionScreen extends Component {
      * @param {function} response callback which takes the data to suggest as argument
      */
     _autocompleteSearch(request, response) {
-        const lcTerm = request.term.toLowerCase();
+        const terms = request.term.toLowerCase().split(" ");
         const limit = 15;
+        const sortLimit = 7;
         const matches = this.state.industries.filter((val) => {
-            return val.label.startsWith(lcTerm);
+            synonymLoop: for (const candidate of [val.label, ...(val.synonyms || '').split(/[|,\n]+/)]) {
+                const trimmed = candidate.trim();
+                for (const term of terms) {
+                    if (!candidate.includes(term)) {
+                        continue synonymLoop;
+                    }
+                }
+                // Note that this will have no effect in languages without spaces.
+                val.wordCount = val.label.split(" ").length;
+                return true;
+            }
+            return false;
         });
-        let results = matches.slice(0, limit);
-        this.labelToId = {};
-        let labels = results.map((val) => val.label);
-        if (labels.length < limit) {
-            let relaxedMatches = this.state.industries.filter((val) => {
-                return val.label.includes(lcTerm) && !labels.includes(val.label);
-            });
-            relaxedMatches = relaxedMatches.slice(0, limit - labels.length);
-            results = results.concat(relaxedMatches);
+        let results = matches;
+        if (matches.length > limit) {
+            // Keep matches with the least number of words so that e.g. "restaurant" remains
+            // available even if there are 15 specific sub-types that have a higher hit count.
+            const matchesPerWordCount = [...matches].sort((x, y) => x.wordCount - y.wordCount);
+            const survivors = matchesPerWordCount.slice(0, limit).map((val) => val.label);
+            results = matches.filter((val) => survivors.includes(val.label));
```

Comment:
It is about keeping survivors in their original order, therefore we need to first establish the list of survivors (1 & 2), then we can filter the correctly sorted values (3).
This is because the priority rule for being a survivor is itself about a different sorting of the values.

Reply:
I was conceptually more concern about the fact that we could filter out too many results or even all the results by doing so? 

Either the (3) (filter out some entries) is useless (as always true and not filtering anything) or it can filter out all the results?

I might be missing something